### PR TITLE
Tui Lazy Attributes and Layout

### PR DESCRIPTION
**Pull request by Demonthos**
Make the Tui renderer lazily update it's attributes, and layouts. To do this it adds a render tree which could also be used in a focus system.

other changes:
1. Events reworked to jump to nodes that are listening to an event instead of traversing the nodes
2. Mouse events work on absolutely positioned elements
3. Adds a native-core crate with generic render tree and Stretch layout utilities
4. Adds a TuiContext that allows the user to gracefully exit (needed for benchmarks)
5. Change alpha channel from f32 to u8 for performance

todo:
- [x] lazy attributes
- [x] lazy layout
- [x] benchmarks
- [x] more documentation
- [x] add tests
- [x] edit reference to reflect changes ([tui](https://github.com/Demonthos/dioxus/blob/lazy_tui/docs/reference/src/platforms/tui.md), [native-core](https://github.com/Demonthos/dioxus/blob/lazy_tui/docs/reference/src/guide/custom_renderer.md))
- [x] pending #https://github.com/DioxusLabs/stretch/pull/2 for layout
- [x] pending #https://github.com/DioxusLabs/dioxus/pull/331 for tests

benchmarks:
updating one element at a time on a square grid (excuding rendering time) 

box count | new | old
-|-|-
9 | 458.84 us | 1.4662 ms                                                                
36 | 2.0414 ms | 19.075 ms                                                                 
81 | 5.9310 ms | 95.541 ms                                                                  
144 | 14.224 ms | 288.39 ms                                                                   
225 | 29.410 ms | 705.10 ms                                                                 
324 | 54.425 ms | 1.6291 s                                                                   
441 | 95.033 ms | 3.0568 s                                                                 
576 | 153.13 ms | 5.3373 s

chart

chart (1)

### CODE REVIEW

**Comment by Demonthos** [comment]
(rebase with master)

**Comment by Demonthos** [comment]
While working on the native renderer there are some improvements I think I can make to the api. I can't re-draft this, but don't merge yet.

- [ ] Multiple of each type of state that can update independently.
- [ ] User-defined state structure that contains all children.
- [ ] State that depends on attributes rather that VNode's (to not expose children and parents).
- [ ] State that depends on state of another type (text size scaling with the width/height).

File: packages/tui/src/layout.rs
Comment (univerz) [inline packages/tui/src/layout.rs:64]: this panics on early return from component (`l.node == None`).
is skipping such children correct?

```rust
let child_layout: Vec<Node> = children.filter_map(|l| l.node).collect();
```

Comment (Demonthos) [reply packages/tui/src/layout.rs:64]: Node should only be none if update_state hasn't been called. The function is resolved from the lowest node in the tree upward, so it always will be some within this function. I should change the Option to an enum with uninitiated to make the code more readable.

Comment (Demonthos) [reply packages/tui/src/layout.rs:64]: It might be worth changing this at the api level so that update takes a mutable reference to a possibly uninitialized self instead of requiring the default trait bounds

Comment (univerz) [reply packages/tui/src/layout.rs:64]: > so it always will be some within this function

i'm not familiar with this codebase yet, just experimenting. minimal example:

```rust
use dioxus::prelude::*;

fn main() {
    dioxus::tui::launch(app)
}

fn app(cx: Scope) -> Element {
    cx.render(rsx! {
        div {
            component()
        }
    })
}

#[inline_props]
fn component(cx: Scope) -> Element {
    None
}
```

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', /home/univerz/.cargo/git/checkouts/dioxus-9d150aea5a02f9fc/3dd18b2/packages/tui/src/layout.rs:64:46
stack backtrace:
   0:     0x55d8526906dd - std::backtrace_rs::backtrace::libunwind::trace::hda39d6860b2d6f70
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x55d8526906dd - std::backtrace_rs::backtrace::trace_unsynchronized::ha1bfdeb508735a0a
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x55d8526906dd - std::sys_common::backtrace::_print_fmt::h4e9e937d66742a36
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x55d8526906dd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h73a948c4553c4f51
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x55d8526b0a9c - core::fmt::write::ha91e2a0e6cdcb7d1
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/fmt/mod.rs:1194:17
   5:     0x55d85268ca31 - std::io::Write::write_fmt::heda409ac36864e96
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/io/mod.rs:1655:15
   6:     0x55d852691fe5 - std::sys_common::backtrace::_print::he8c3809f69ea2342
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x55d852691fe5 - std::sys_common::backtrace::print::hecbd7b926628aa09
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x55d852691fe5 - std::panicking::default_hook::{{closure}}::h8fbf7884094b6f2e
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:295:22
   9:     0x55d852691c59 - std::panicking::default_hook::h6334d4a28312e99b
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:314:9
  10:     0x55d852692692 - std::panicking::rust_panic_with_hook::h73dbf64f0060aa58
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:698:17
  11:     0x55d852692539 - std::panicking::begin_panic_handler::{{closure}}::h57f62f8b0c4aa3d9
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:586:13
  12:     0x55d852690b94 - std::sys_common::backtrace::__rust_end_short_backtrace::h3194a80099ffb120
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:138:18
  13:     0x55d8526922a9 - rust_begin_unwind
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:584:5
  14:     0x55d8526afab3 - core::panicking::panic_fmt::h77eb0c9a0a627689
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/panicking.rs:142:14
  15:     0x55d8526af8fd - core::panicking::panic::h72eee331f65d5eb5
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/panicking.rs:48:5
  16:     0x55d852601c89 - core::option::Option<T>::unwrap::hda627f64fff24639
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/option.rs:755:21
  17:     0x55d852601c89 - <dioxus_tui::layout::StretchLayout as dioxus_native_core::real_dom::BubbledUpState>::reduce::h1bc15ed5d5acd0cd
                               at /home/univerz/.cargo/git/checkouts/dioxus-9d150aea5a02f9fc/3dd18b2/packages/tui/src/layout.rs:64:39
  18:     0x55d8525f7b7c - dioxus_native_core::real_dom::RealDom<US,DS>::update_state::h4d5583689fc8dae3
                               at /home/univerz/.cargo/git/checkouts/dioxus-9d150aea5a02f9fc/3dd18b2/packages/native-core/src/real_dom.rs:222:13
  19:     0x55d85260208f - dioxus_tui::launch_cfg::h8208bacb9ec9fc1b
                               at /home/univerz/.cargo/git/checkouts/dioxus-9d150aea5a02f9fc/3dd18b2/packages/tui/src/lib.rs:77:24
  20:     0x55d8525e2133 - core::ops::function::FnOnce::call_once::h77b6b1956a25dc65
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/ops/function.rs:248:5
  21:     0x55d8525e2133 - std::sys_common::backtrace::__rust_begin_short_backtrace::h65a6c301e7e033ab
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/sys_common/backtrace.rs:122:18
  22:     0x55d8525e2149 - std::rt::lang_start::{{closure}}::h4fe2d00309e56e16
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/rt.rs:145:18
  23:     0x55d852687c0e - core::ops::function::impls::<impl core::ops::function::FnOnce<A> for &F>::call_once::h8747dac316cfe8b1
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/core/src/ops/function.rs:280:13
  24:     0x55d852687c0e - std::panicking::try::do_call::ha8cfd36c99c35841
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:492:40
  25:     0x55d852687c0e - std::panicking::try::h44599ca0771bc4e2
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:456:19
  26:     0x55d852687c0e - std::panic::catch_unwind::hfc289307f73cb8b7
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panic.rs:137:14
  27:     0x55d852687c0e - std::rt::lang_start_internal::{{closure}}::h9fadec4e4222aa5a
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/rt.rs:128:48
  28:     0x55d852687c0e - std::panicking::try::do_call::h742ca90b47c7614a
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:492:40
  29:     0x55d852687c0e - std::panicking::try::ha8060cd5fe01b8e1
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panicking.rs:456:19
  30:     0x55d852687c0e - std::panic::catch_unwind::h3548704cd9c5e356
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/panic.rs:137:14
  31:     0x55d852687c0e - std::rt::lang_start_internal::hc6526c5151dc66ec
                               at /rustc/90ca44752a79dd414d9a0ccf7a74533a99080988/library/std/src/rt.rs:128:20
  32:     0x55d8525e2298 - main
  33:     0x7ff8d888034a - <unknown>
  34:     0x7ff8d88803fc - __libc_start_main
  35:     0x55d8525e2061 - _start
  36:                0x0 - <unknown>
```

Comment (Demonthos) [reply packages/tui/src/layout.rs:64]: Ok, It looks like it is creating a placeholder node for the component, which is not handled in the match statement.

Comment (Demonthos) [reply packages/tui/src/layout.rs]: This is fixed in the latest version.

File: docs/reference/src/guide/custom_renderer.md
Comment (Demonthos) [inline docs/reference/src/guide/custom_renderer.md:1]: Updated from [docs/guide/advanced-guides](https://github.com/DioxusLabs/dioxus/blob/master/docs/guide/src/advanced-guides/custom-renderer.md)

File: packages/tui/src/hooks.rs
Comment (Demonthos) [inline packages/tui/src/hooks.rs:368]: This only ran when work with deadline was called which was causing a race condition.

**Comment by Demonthos** [comment]
I would love to have nodes that depend on mutable dependencies, but without GATs I don't think it is possible (same as the borrowed iterator problem).

**Comment by jkelleyrtp** [comment]
I wonder if there's a way we can merge this piece-by-piece, or maybe try pair-reviewing it to make the process go a bit faster? I'm going through the changes locally and am really impressed :)

**Comment by Demonthos** [comment]
I can pull out the tui bugfixes and quality of life improvements. I would be happy to pair on it!

**Comment by Demonthos** [comment]
I didn't realize there were any attributes missing; I think that is last one without leaking children and parents.

**Comment by Demonthos** [comment]
Nothing changed in the force-push. I just committed to the wrong branch and reverted it.

**Comment by jkelleyrtp** [comment]
Is this good to review? If so I'll see if I can get it in today.

**Comment by Demonthos** [comment]
Yes

**Comment by Demonthos** [comment]
Sorry to change things while you are reviewing, but I found a bug. Nodes without any attributes effecting layout don't get the layout initialized causing a panic. This should fix that.

**Review by jkelleyrtp — APPROVED** [review]

**Comment by jkelleyrtp** [comment]
There's a lot going into this PR but not much of it touches the broader workspace.

I need some more time to dig into native-core - seems like some of the widget system seems pretty verbose/complex. Though, I imagine the frequency in which people should be writing widgets to be pretty low, so there's no issue with making it somewhat obtuse.

I added a few comments, but for the most part, native-core-macro and native-core get a rubber stamp of approval. I'm not sure what your timeline is on wanting to publish/release these crates.

File: packages/tui/src/style_attributes.rs
Comment (Demonthos) [reply packages/tui/src/style_attributes.rs]: Oops. Tried to parse these with regex. 😂

**Pull request merged**
